taichi/backends/cuda/codegen_cuda.cpp

```diff
@@ -240,6 +240,48 @@ class CodeGenLLVMCUDA : public CodeGenLLVM {
 #undef UNARY_STD
   }
 
+  // Not all reduction statements can be optimized.
+  // If the operation cannot be optimized, this function returns nullptr.
+  llvm::Value *optimized_reduction(AtomicOpStmt *stmt) {
+    if (!stmt->is_reduction) {
+      return nullptr;
+    }
+    if (!stmt->val->ret_type->is<PrimitiveType>()) {
```

Comment:
nit: I believe this deserves a `TI_ASSERT`, because if this is not a primitive, it's a bug in our code.

Reply:
Sure. That makes sense.